File: Jogo da Velha/Jogo da Velha.kt

// MainActivity.kt
package com.exemplo.jogodavelha

import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.gridlayout.widget.GridLayout

class MainActivity : AppCompatActivity() {
    
    private lateinit var gameLogic: JogoDaVelhaLogic
    private lateinit var statusText: TextView
    private lateinit var resetButton: Button
    private lateinit var gridLayout: GridLayout
    private val buttons = Array(3) { Array<Button?>(3) { null } }
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        
        initializeViews()
        initializeGame()
        setupClickListeners()
    }
    
    private fun initializeViews() {
        statusText = findViewById(R.id.statusText)
        resetButton = findViewById(R.id.resetButton)
        gridLayout = findViewById(R.id.gridLayout)
        
        // Inicializa os botões do tabuleiro
        for (i in 0..2) {
            for (j in 0..2) {
                val buttonId = resources.getIdentifier("btn_${i}_${j}", "id", packageName)
                buttons[i][j] = findViewById(buttonId)
            }
        }
    }
    
    private fun initializeGame() {
        gameLogic = JogoDaVelhaLogic()
        updateUI()
    }
    
    private fun setupClickListeners() {
        // Listener para cada botão do tabuleiro
        for (i in 0..2) {
            for (j in 0..2) {
                buttons[i][j]?.setOnClickListener {
                    onCellClicked(i, j)
                }
            }
        }
        
        // Listener para botão de reset
        resetButton.setOnClickListener {
            resetGame()
        }
    }
    
    private fun onCellClicked(linha: Int, coluna: Int) {
        if (gameLogic.fazerJogada(linha, coluna)) {
            updateUI()
            
            when (gameLogic.obterEstadoJogo()) {
                is EstadoJogo.Vencedor -> {
                    showWinnerMessage(gameLogic.obterEstadoJogo() as EstadoJogo.Vencedor)
                }
                is EstadoJogo.Empate -> {
                    showDrawMessage()
                }
                else -> {
                    // Jogo continua
                }
            }
        } else {
            Toast.makeText(this, "Jogada inválida!", Toast.LENGTH_SHORT).show()
        }
    }
    
    private fun updateUI() {
        // Atualiza o tabuleiro visual
        for (i in 0..2) {
            for (j in 0..2) {
                val cellValue = gameLogic.obterCelula(i, j)
                buttons[i][j]?.text = when (cellValue) {
                    Jogador.X -> "X"
                    Jogador.O -> "O"
                    else -> ""
                }
                
                // Desabilita botão se célula ocupada ou jogo terminado
                buttons[i][j]?.isEnabled = cellValue == null && gameLogic.jogoAtivo
            }
        }
        
        // Atualiza texto de status
        statusText.text = when (val estadoJogo = gameLogic.obterEstadoJogo()) {
            is EstadoJogo.JogadorAtual -> "Vez do jogador ${estadoJogo.jogador.symbol}"
            is EstadoJogo.Vencedor -> "Jogador ${estadoJogo.jogador.symbol} venceu!"
            is EstadoJogo.Empate -> "Empate!"
        }
    }
    
    private fun showWinnerMessage(vencedor: EstadoJogo.Vencedor) {
        Toast.makeText(
            this, 
            "🎉 Jogador ${vencedor.jogador.symbol} venceu!", 
            Toast.LENGTH_LONG
        ).show()
    }
    
    private fun showDrawMessage() {
        Toast.makeText(this, "🤝 Empate!", Toast.LENGTH_LONG).show()
    }
    
    private fun resetGame() {
        gameLogic.reiniciarJogo()
        updateUI()
    }
}

// JogoDaVelhaLogic.kt - Lógica do jogo adaptada para Kotlin
enum class Jogador(val symbol: String) {
    X("X"),
    O("O");
    
    fun outro(): Jogador = if (this == X) O else X
}

sealed class EstadoJogo {
    data class JogadorAtual(val jogador: Jogador) : EstadoJogo()
    data class Vencedor(val jogador: Jogador) : EstadoJogo()
    object Empate : EstadoJogo()
}

class JogoDaVelhaLogic {
    private val tabuleiro = Array(3) { Array<Jogador?>(3) { null } }
    private var jogadorAtual = Jogador.X
    var jogoAtivo = true
        private set
    private var vencedor: Jogador? = null
    
    fun fazerJogada(linha: Int, coluna: Int): Boolean {
        if (!jogoAtivo || !posicaoValida(linha, coluna)) {
            return false
        }
        
        // Faz a jogada
        tabuleiro[linha][coluna] = jogadorAtual
        
        // Verifica resultado
        when {
            verificarVitoria() -> {
                vencedor = jogadorAtual
                jogoAtivo = false
            }
            tabuleiroCompleto() -> {
                jogoAtivo = false
            }
            else -> {
                jogadorAtual = jogadorAtual.outro()
            }
        }
        
        return true
    }
    
    fun obterCelula(linha: Int, coluna: Int): Jogador? {
        return if (linha in 0..2 && coluna in 0..2) {
            tabuleiro[linha][coluna]
        } else null
    }
    
    fun obterEstadoJogo(): EstadoJogo {
        return when {
            vencedor != null -> EstadoJogo.Vencedor(vencedor!!)
            !jogoAtivo -> EstadoJogo.Empate
            else -> EstadoJogo.JogadorAtual(jogadorAtual)
        }
    }
    
    fun reiniciarJogo() {
        for (i in 0..2) {
            for (j in 0..2) {
                tabuleiro[i][j] = null
            }
        }
        jogadorAtual = Jogador.X
        jogoAtivo = true
        vencedor = null
    }
    
    private fun posicaoValida(linha: Int, coluna: Int): Boolean {
        return linha in 0..2 && 
               coluna in 0..2 && 
               tabuleiro[linha][coluna] == null
    }
    
    private fun verificarVitoria(): Boolean {
        return verificarLinhas() || verificarColunas() || verificarDiagonais()
    }
    
    private fun verificarLinhas(): Boolean {
        for (linha in 0..2) {
            if (tabuleiro[linha][0] != null &&
                tabuleiro[linha][0] == tabuleiro[linha][1] &&
                tabuleiro[linha][1] == tabuleiro[linha][2]) {
                return true
            }
        }
        return false
    }
    
    private fun verificarColunas(): Boolean {
        for (coluna in 0..2) {
            if (tabuleiro[0][coluna] != null &&
                tabuleiro[0][coluna] == tabuleiro[1][coluna] &&
                tabuleiro[1][coluna] == tabuleiro[2][coluna]) {
                return true
            }
        }
        return false
    }
    
    private fun verificarDiagonais(): Boolean {
        // Diagonal principal
        if (tabuleiro[0][0] != null &&
            tabuleiro[0][0] == tabuleiro[1][1] &&
            tabuleiro[1][1] == tabuleiro[2][2]) {
            return true
        }
        
        // Diagonal secundária
        if (tabuleiro[0][2] != null &&
            tabuleiro[0][2] == tabuleiro[1][1] &&
            tabuleiro[1][1] == tabuleiro[2][0]) {
            return true
        }
        
        return false
    }
    
    private fun tabuleiroCompleto(): Boolean {
        for (linha in 0..2) {
            for (coluna in 0..2) {
                if (tabuleiro[linha][coluna] == null) {
                    return false
                }
            }
        }
        return true
    }
}

File: Jogo da Velha/Jogo da Velha.py

class JogoDaVelha:
    def __init__(self):
        # Tabuleiro 3x3 representado como lista de listas
        # ' ' = vazio, 'X' = jogador 1, 'O' = jogador 2
        self.tabuleiro = [[' ' for _ in range(3)] for _ in range(3)]
        self.jogador_atual = 'X'  # X sempre começa
        self.jogo_ativo = True
        self.vencedor = None
    
    def exibir_tabuleiro(self):
        """Exibe o tabuleiro atual no console"""
        print("\n  0   1   2")
        for i in range(3):
            print(f"{i} {self.tabuleiro[i][0]} | {self.tabuleiro[i][1]} | {self.tabuleiro[i][2]}")
            if i < 2:
                print("  ---------")
    
    def posicao_valida(self, linha, coluna):
        """Verifica se a posição é válida e está vazia"""
        if linha < 0 or linha > 2 or coluna < 0 or coluna > 2:
            return False
        return self.tabuleiro[linha][coluna] == ' '
    
    def fazer_jogada(self, linha, coluna):
        """Executa uma jogada se for válida"""
        if not self.jogo_ativo:
            return False
        
        if not self.posicao_valida(linha, coluna):
            return False
        
        # Coloca o símbolo do jogador atual
        self.tabuleiro[linha][coluna] = self.jogador_atual
        
        # Verifica se há vencedor após a jogada
        if self.verificar_vitoria():
            self.vencedor = self.jogador_atual
            self.jogo_ativo = False
        elif self.tabuleiro_cheio():
            # Empate
            self.jogo_ativo = False
        else:
            # Troca o jogador
            self.trocar_jogador()
        
        return True
    
    def verificar_vitoria(self):
        """Verifica todas as condições de vitória"""
        return (self.verificar_linhas() or 
                self.verificar_colunas() or 
                self.verificar_diagonais())
    
    def verificar_linhas(self):
        """Verifica vitória nas linhas horizontais"""
        for linha in self.tabuleiro:
            if linha[0] == linha[1] == linha[2] != ' ':
                return True
        return False
    
    def verificar_colunas(self):
        """Verifica vitória nas colunas verticais"""
        for col in range(3):
            if (self.tabuleiro[0][col] == 
                self.tabuleiro[1][col] == 
                self.tabuleiro[2][col] != ' '):
                return True
        return False
    
    def verificar_diagonais(self):
        """Verifica vitória nas diagonais"""
        # Diagonal principal (top-left para bottom-right)
        if (self.tabuleiro[0][0] == 
            self.tabuleiro[1][1] == 
            self.tabuleiro[2][2] != ' '):
            return True
        
        # Diagonal secundária (top-right para bottom-left)
        if (self.tabuleiro[0][2] == 
            self.tabuleiro[1][1] == 
            self.tabuleiro[2][0] != ' '):
            return True
        
        return False
    
    def tabuleiro_cheio(self):
        """Verifica se o tabuleiro está completamente preenchido"""
        for linha in self.tabuleiro:
            for celula in linha:
                if celula == ' ':
                    return False
        return True
    
    def trocar_jogador(self):
        """Alterna entre os jogadores X e O"""
        self.jogador_atual = 'O' if self.jogador_atual == 'X' else 'X'
    
    def reiniciar_jogo(self):
        """Reinicia o jogo para o estado inicial"""
        self.tabuleiro = [[' ' for _ in range(3)] for _ in range(3)]
        self.jogador_atual = 'X'
        self.jogo_ativo = True
        self.vencedor = None
    
    def obter_estado_jogo(self):
        """Retorna o estado atual do jogo"""
        if not self.jogo_ativo:
            if self.vencedor:
                return f"Jogador {self.vencedor} venceu!"
            else:
                return "Empate!"
        else:
            return f"Vez do jogador {self.jogador_atual}"


# Exemplo de uso e teste da lógica
def main():
    jogo = JogoDaVelha()
    
    print("=== JOGO DA VELHA ===")
    print("Digite as coordenadas no formato: linha coluna (0-2)")
    print("Exemplo: 1 1 para o centro do tabuleiro")
    
    while jogo.jogo_ativo:
        jogo.exibir_tabuleiro()
        print(f"\n{jogo.obter_estado_jogo()}")
        
        try:
            entrada = input("Digite sua jogada (linha coluna): ").split()
            if len(entrada) != 2:
                print("Formato inválido! Use: linha coluna")
                continue
                
            linha = int(entrada[0])
            coluna = int(entrada[1])
            
            if jogo.fazer_jogada(linha, coluna):
                # Jogada válida executada
                pass
            else:
                print("Jogada inválida! Tente novamente.")
                
        except ValueError:
            print("Digite apenas números!")
        except KeyboardInterrupt:
            print("\nJogo interrompido!")
            break
    
    # Resultado final
    jogo.exibir_tabuleiro()
    print(f"\n{jogo.obter_estado_jogo()}")


if __name__ == "__main__":
    main()
    